gift-certificate-data/src/main/java/com/epam/esm/giftcertificatemodule3/dao/impl/GiftCertificateDAOHibernate.java

```diff
@@ -2,54 +2,142 @@
 
 import com.epam.esm.giftcertificatemodule3.dao.GiftCertificateDAO;
 import com.epam.esm.giftcertificatemodule3.entity.GiftCertificate;
-import org.hibernate.Session;
-import org.hibernate.query.Query;
+import com.epam.esm.giftcertificatemodule3.entity.Tag;
+import com.epam.esm.giftcertificatemodule3.model.SearchParametersHolder;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.stereotype.Repository;
 
 import javax.persistence.EntityManager;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.Query;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.*;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
 import java.util.List;
 
 @Repository
 public class GiftCertificateDAOHibernate implements GiftCertificateDAO {
 
-    private EntityManager entityManager;
+    private final EntityManagerFactory emf;
```

Comment:
Factory is not needed here. Just inject `EntityManager`. It should already be request scoped proxy.